File: src/main/java/io/binac/leetcode/DecodeWays.java
package io.binac.leetcode;

/**
 * A message containing letters from A-Z is being encoded to numbers using the following mapping:
 * <blockquote><pre>
 *     'A' -> 1
 *     'B' -> 2
 *     ...
 *     'Z' -> 26
 * </blockquote></pre>
 * Given a non-empty string containing only digits, determine the total number of ways to decode it.
 * <p>
 * <p>Example 1:
 * <blockquote><pre>
 *     Input: "12"
 *     Output: 2
 *     Explanation: It could be decoded as "AB" (1 2) or "L" (12).
 * </blockquote></pre>
 * Example 2:
 * <blockquote><pre>
 *     Input: "226"
 *     Output: 3
 *     Explanation: It could be decoded as "BZ" (2 26), "VF" (22 6), or "BBF" (2 2 6).
 * </blockquote></pre>
 */
public class DecodeWays {
    public static class Solution1 {
        public int numDecodings(String s) {
            final int len = s.length();
            if (len == 0)
                throw new IllegalArgumentException("s can't be empty");

            int pre = 1, cur = s.charAt(0) == '0' ? 0 : 1, tmp;
            for (int i = 1; i < len && cur != 0; ++i) {
                tmp = cur;
                if (s.charAt(i - 1) == '1' || (s.charAt(i - 1) == '2' && s.charAt(i) <= '6')) {
                    if (s.charAt(i) == '0')
                        cur = pre;
                    else
                        cur += pre;
                } else if (s.charAt(i) == '0') {
                    cur = 0;
                }
                pre = tmp;
            }
            return cur;
        }
    }
}

File: python/leetcodepy/decode_ways.py
"""
A message containing letters from A-Z is being encoded to numbers using the following mapping:

'A' -> 1
'B' -> 2
...
'Z' -> 26

Given a non-empty string containing only digits, determine the total number of ways to decode it.

Example 1:

Input: "12"
Output: 2
Explanation: It could be decoded as "AB" (1 2) or "L" (12).

Example 2:

Input: "226"
Output: 3
Explanation: It could be decoded as "BZ" (2 26), "VF" (22 6), or "BBF" (2 2 6).
"""


class Solution1:
    def numDecodings(self, s):
        """
        :type s: str
        :rtype: int
        """
        if not s:
            raise ValueError
        size = len(s)
        dp = [0] * size
        if s[0] != '0':
            dp[0] = 1
        if size == 1:
            return dp[0]
        if s[1] == '0':
            if 1 <= int(s[0]) <= 2:
                dp[1] = 1
        else:
            dp[1] = dp[0]
            if s[0] != '0' and int(s[:2]) <= 26:
                dp[1] += 1
        if dp[0] == 0 or dp[1] == 0:
            return 0
        for i in range(2, size):
            if s[i] == '0':
                if 1 <= int(s[i - 1]) <= 2:
                    dp[i] = dp[i - 2]
            else:
                dp[i] = dp[i - 1]
                if s[i - 1] != '0' and int(s[i - 1:i + 1]) <= 26:
                    dp[i] += dp[i - 2]
            if dp[i] == 0:
                return 0
        return dp[size - 1]
